src/poly/mod.rs

```diff
@@ -61,27 +111,76 @@ impl<F, V, M> Expr<F, V, M> {
+        match self {
+            Expr::Const(v, m) => Expr::Const(v.clone(), f(m)),
+            Expr::Sum(ses, m) => Expr::Sum(ses.iter().map(|e| e.map_meta(&f)).collect(), f(m)),
+            Expr::Mul(ses, m) => Expr::Mul(ses.iter().map(|e| e.map_meta(&f)).collect(), f(m)),
+            Expr::Neg(se, m) => Expr::Neg(Box::new(se.map_meta(&f)), f(m)),
+            Expr::Pow(se, exp, m) => Expr::Pow(Box::new(se.map_meta(&f)), *exp, f(m)),
+            Expr::Query(v, m) => Expr::Query(v.clone(), f(m)),
+            Expr::Halo2Expr(e, m) => Expr::Halo2Expr(e.clone(), f(m)),
+            Expr::MI(se, m) => Expr::MI(Box::new(se.map_meta(&f)), f(m)),
+        }
+    }
 }
 
-impl<F: Field + Hash, V: Debug + Clone + Eq + Hash> Expr<F, V, ()> {
-    /// Uses Schwartz-Zippel Lemma to hash
-    pub fn hash(&self, assignments: &VarAssignments<F, V>) -> Expr<F, V, HashResult> {
-        let mut hasher = DefaultHasher::new();
+impl<F: Field + Hash, V: Debug + Clone + Eq + Hash, M: Meta> Expr<F, V, M> {
+    pub fn with_meta<N: Meta>(&self, assignments: &VarAssignments<F, V>) -> Expr<F, V, N> {
+        let expr_without_meta = self.without_meta();
+        let new_meta = N::from_expr(&expr_without_meta, assignments);
+        match self {
+            Expr::Const(v, _) => Expr::Const(*v, new_meta),
+            Expr::Sum(ses, _) => Expr::Sum(
+                ses.iter().map(|e| e.with_meta(assignments)).collect(),
+                new_meta,
+            ),
+            Expr::Mul(ses, _) => Expr::Mul(
+                ses.iter().map(|e| e.with_meta(assignments)).collect(),
+                new_meta,
+            ),
+            Expr::Neg(se, _) => Expr::Neg(Box::new(se.with_meta(assignments)), new_meta),
+            Expr::Pow(se, exp, _) => Expr::Pow(Box::new(se.with_meta(assignments)), *exp, new_meta),
+            Expr::Query(v, _) => Expr::Query(v.clone(), new_meta),
+            Expr::Halo2Expr(e, _) => Expr::Halo2Expr(e.clone(), new_meta),
+            Expr::MI(se, _) => Expr::MI(Box::new(se.with_meta(assignments)), new_meta),
+        }
+    }
 
-        if let Some(result) = self.eval(assignments) {
-            result.hash(&mut hasher);
+    pub fn apply_subexpressions<T>(&self, mut f: T) -> Self
+    where
+        T: FnMut(&Self) -> Self,
+    {
+        match self {
+            Expr::Sum(ses, m) => Expr::Sum(ses.iter().map(&mut f).collect(), m.clone()),
+            Expr::Mul(ses, m) => Expr::Mul(ses.iter().map(&mut f).collect(), m.clone()),
+            Expr::Neg(se, m) => Expr::Neg(Box::new(f(se)), m.clone()),
+            Expr::Pow(se, exp, m) => Expr::Pow(Box::new(f(se)), *exp, m.clone()),
+            Expr::MI(se, m) => Expr::MI(Box::new(f(se)), m.clone()),
+            _ => self.clone(),
         }
+    }
+}
 
-        let hash_result = HashResult {
-            hash: hasher.finish(),
-            degree: self.degree(),
-        };
+impl<F: Field, V: Clone, M: Meta> Expr<F, V, M> {
```

Comment:
@rutefig why does M has to be of Meta for this method to work? We should not add traits to the types if they are not necessary